### PR TITLE
Add MultiHead Attention

### PR DESCRIPTION
**Pull request by crcrpar**
After reading 
    - [cudnn](https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnMultiHeadAttnForward)
    - [pytorch/fairseq](https://github.com/pytorch/fairseq/blob/master/fairseq/modules/multihead_attention.py) (the base impl. of [pytorch/pytorch](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/activation.py#L682))
    - [dmlc/gluon-nlp](https://github.com/dmlc/gluon-nlp/blob/master/src/gluonnlp/model/attention_cell.py)

I chose pytorch's impl as a reference.

**TODO**
- [x] test
- [x] documentation

Since I'm not familiar with Attention enough, any comment is highly welcomed.

### CODE REVIEW

File: chainer/links/connection/multihead_attention.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
+import typing as tp  # NOQA
```
Comment (himkt) [inline chainer/links/connection/multihead_attention.py]: It seems not to be used.

File: chainer/links/connection/multihead_attention.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+import numpy
+
+from chainer import backend
+from chainer import functions
+from chainer import initializers
+from chainer import link
+from chainer import links
+from chainer import types  # NOQA
```
Comment (himkt) [inline chainer/links/connection/multihead_attention.py]: Ditto.

File: chainer/functions/connection/multihead_attention.py [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    def _in_proj_v(value):
+        return _in_proj(value, start=2 * embedding_size, weight_idx=2)
+
+    if embedding_size % n_head != 0:
+        raise ValueError(
+            '`embedding_size` ({}) need to be '.format(embedding_size) +
+            'divisible by `n_head` ({})'.format(embedding_size, n_head))
+    if (bias_k is None) != (bias_v is None):
+        raise ValueError
+    qkv_same = (query is key) and (query is value)
```
Comment (crcrpar) [inline chainer/functions/connection/multihead_attention.py]: Can I handle the case where key and value are None as self-attention?

Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: e.g.

```python
qkv_same = ((query is key) and (query is value)) or (key is None and query is None)
```

File: chainer/links/connection/multihead_attention.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+    def __init__(
+        self,
+        n_head,                 # type: int
+        embedding_size,             # type: int
+        self_attention=False,   # type: tp.Optional[bool]
+        ksize=None,             # type: tp.Optional[int]
+        vsize=None,             # type: tp.Optional[int]
+        attention_dropout=0.0,  # type: tp.Optional[float]
+        post_dropout=0.0,       # type: tp.Optional[float]
```
Comment (crcrpar) [inline chainer/links/connection/multihead_attention.py]: This value is not used as of now but necessary for cuDNN.
ref: https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#cudnnSetAttnDescriptor

File: chainer/functions/connection/multihead_attention.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+            A batch of key vectors whose shape is :math:`(S, B, E)` where
+            :math:`S` is the source sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
+        value (:class:`~chainer.Variable` or :ref:`ndarray`)
+            A batch of value vectors whose shape is :math:`(S, B, E)` where
+            :math:`S` is the source sequence length, :math:`B` is the
+            batch size, and :math:`E` is the embedding size.
+        expected_embedding_size (int): Total number of units of the model.
+        n_head (int): The number of parallel attention heads.
+        proj_in_W (tuple of :class:`~chainer.Variable`\\s, :class:`~chainer.Variable` or :ref:`ndarray`):
```
Comment (crcrpar) [inline chainer/functions/connection/multihead_attention.py]: How can fix this line to obey pep8?

**Comment by crcrpar** [comment]
Now I'm done with porting pytorch implementation.

@himkt @keisuke-umezawa Could you skim this?

File: chainer/links/connection/multihead_attention.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+                self.bias_k = variable.Parameter(
+                    _initial_bias, (self.embedding_size,))  # type: variable.Variable  # NOQA
+                self.bias_v = variable.Parameter(
+                    _initial_bias, (self.embedding_size,))  # type: variable.Variable  # NOQA
+            else:
+                self.bias_k, self.bias_v = None, None
+
+    def forward(
+        self,
+        query,                     # type: tp.Optional[InputType]
```
Comment (crcrpar) [inline chainer/links/connection/multihead_attention.py]:
```suggestion
        query,                     # type: InputType
```

File: chainer/links/connection/multihead_attention.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            return_weights (bool):
+                If ``True``, return both attention and attention weights.
+        Returns:
+            tuple of :class:`~chainer.Variable`\\ s:
+                if `return_weights` is ``True``.
+                Otherwise, :class:`~chainer.Variable`.
+                The first element is context vector shaped :math:`(L, B, E)`
+                the second is attention weights shaped :math:`(B, L, S)`.
+        """
+        # TODO (crcrpar): Support cuDNN MultiHeadAttn when CuPy supports it.
```
Comment (crcrpar) [inline chainer/links/connection/multihead_attention.py]: Move this line to chainer/functions/connection/multihead_attention.py

File: chainer/functions/connection/multihead_attention.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+    proj_out_b=None,  # type: variable.Variable
+    add_zero_attn=False,  # type: bool
+    dropout=0,  # type: float
+    post_dropout=0,  # type: float
+    key_padding_mask=None,  # type: tp.Optional[InputType]
+    attn_mask=None,  # type: tp.Optional[InputType]
+    scaling=None,  # type: tp.Optional[float]
+    return_weights=True  # type: tp.Optional[bool]
+):
+    # type: (...) -> tp.Union[variable.Variable, tp.Tuple[variable.Variable, variable.Variable]]  # NOQA
```
Comment (crcrpar) [inline chainer/functions/connection/multihead_attention.py]:
```suggestion
    # type: (...) -> tp.Tuple[variable.Variable, variable.Variable]  # NOQA
```

File: chainer/links/connection/multihead_attention.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+            key (:class:`~chainer.Variable` or :ref:`ndarray`):
+                The key of the memory vectors with the shape of
+                (time, batch_size, key_in_size).
+            value (:class:`~chainer.Variable` or :ref:`ndarray`):
+                The value of the memory with the shape of
+                (time, batch_size, value_in_size).
+            key_padding_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+                If not ``None``, mask the memory slots.
+                The shape is (batch_size, source_length).
+                Each value is 0 or 1 where 1 means that
```
Comment (crcrpar) [inline chainer/links/connection/multihead_attention.py]:
```suggestion
                Each value is ``False`` or ``True`` where ``True`` means that
```

File: chainer/functions/connection/multihead_attention.py [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+    key,  # type: InputType
+    value,  # type: InputType
+    proj_in_W,  # type: tp.Union[variable.Variable, tp.Tuple[variable.Variable, variable.Variable, variable.Variable]]  # NOQA
+    proj_in_b=None,  # type: variable.Variable
+    bias_k=None,  # type: variable.Variable
+    bias_v=None,  # type: variable.Variable
+    proj_out_W=None,  # type: variable.Variable
+    proj_out_b=None,  # type: variable.Variable
+    add_zero_attn=False,  # type: bool
+    dropout=0,  # type: float
```
Comment (crcrpar) [inline chainer/functions/connection/multihead_attention.py]:
```suggestion
    attention_dropout=0,  # type: float
```
Use the name used in `L.MultiHeadAttention` and fix `from chainer import dropout as dropout_func` to `from chainer.functions import dropout`

File: chainer/functions/connection/multihead_attention.py [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+            will be ignored by the attention.
+            The shape is :math:`(B, S)` where :math:`B` is the batch size,
+            and :math:`S` is the source sequence length.
+        attn_mask (:class:`~chainer.Variable` or :ref:`ndarray`):
+            Mask help attention ignores certain positions.
+            The shape is :math:`(L, L)` where :math:`L` is
+            the target sequence length.
+        return_weights (bool): If ``True``, return averaged attention weights.
+
+    Outputs:
```
Comment (crcrpar) [inline chainer/functions/connection/multihead_attention.py]: Fix this by mimicking https://github.com/chainer/chainer/blob/master/chainer/functions/connection/n_step_rnn.py#L443

> 
    Returns:
        tuple: This function returns a tuple containing three elements,
        ``hy`` and ``ys``.
        - ``hy`` is an updated hidden states whose shape is same as ``hx``.
        - ``ys`` is a list of :class:`~chainer.Variable` . Each element
          ``ys[t]`` holds hidden states of the last layer corresponding
          to an input ``xs[t]``. Its shape is ``(B_t, N)`` where ``B_t`` is
          mini-batch size for time ``t``, and ``N`` is size of hidden
          units. Note that ``B_t`` is the same value as ``xs[t]``.

File: chainer/links/connection/multihead_attention.py [hunk 12]
```diff
@@ -0,0 +1,198 @@
...
+                the memory slot will not be used.
+            attention_mask (:class:`~chainer.Variable` or :ref:`ndarray`)
+            return_weights (bool):
+                If ``True``, return both attention and attention weights.
+        Returns:
```
Comment (crcrpar) [inline chainer/links/connection/multihead_attention.py]: Fix this by mimicking https://github.com/chainer/chainer/blob/master/chainer/functions/connection/n_step_rnn.py#L443

> 
    Returns:
        tuple: This function returns a tuple containing three elements,
        ``hy`` and ``ys``.
        - ``hy`` is an updated hidden states whose shape is same as ``hx``.
        - ``ys`` is a list of :class:`~chainer.Variable` . Each element
          ``ys[t]`` holds hidden states of the last layer corresponding
          to an input ``xs[t]``. Its shape is ``(B_t, N)`` where ``B_t`` is
          mini-batch size for time ``t``, and ``N`` is size of hidden
          units. Note that ``B_t`` is the same value as ``xs[t]``.

File: chainer/functions/connection/multihead_attention.py [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+from chainer.functions.array import split_axis
+from chainer.functions.array import repeat
+from chainer.functions.array import reshape
+from chainer.functions.array import transpose
+from chainer.functions.array import where
+from chainer.functions.math import average
+from chainer.functions.math import matmul
+from chainer.functions.noise import dropout
+from chainer.functions.connection import linear
+from chainer import types  # NOQA
```
Comment (himkt) [inline chainer/functions/connection/multihead_attention.py]: It seems that NOQA is no more required! :beer:

File: chainer/links/connection/multihead_attention.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
+import typing as tp  # NOQA
```
Comment (himkt) [inline chainer/links/connection/multihead_attention.py]: ditto.

File: chainer/links/connection/multihead_attention.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+
+
+from chainer import functions
+from chainer import initializers
+from chainer import link
+from chainer import links
+from chainer import types  # NOQA
```
Comment (himkt) [inline chainer/links/connection/multihead_attention.py]: ditto.

File: chainer/functions/connection/multihead_attention.py (hunk 13 above)
Comment (crcrpar) [reply chainer/functions/connection/multihead_attention.py]: Thank you 👍

**Comment by himkt** [comment]
Sorry for being late, we want to merge this PR as the experimental functionality
before merging https://github.com/cupy/cupy/pull/2215.
So, could do please resolve conflicts and add a statement to make MHA modules experimental?

example: https://github.com/chainer/chainer/blob/721986ebfdf9375f52ef71f885ad8ea6402a5a03/chainermn/links/n_step_rnn.py#L84

ref. https://docs.chainer.org/en/stable/reference/generated/chainer.utils.experimental.html

**Comment by crcrpar** [comment]
Could you elaborate on the motivation behind experimental?

**Comment by crcrpar** [comment]
Ah, I made a big mistake in resolving conflicts.

**Comment by crcrpar** [comment]
workin' on fixing this mistake.

**Comment by 2sin18** [comment]
Why this pull request is closed? What's the problem of using cudnn for multihead attention?

**Comment by crcrpar** [comment]
@2sin18 Chainer has been in maintenance, not actively developed anymore. So they and I decided not to make this pull request & #7879 merged. Thank you for your interest and sorry not to have made this happen.
